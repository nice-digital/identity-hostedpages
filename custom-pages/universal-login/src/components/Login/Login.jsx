import React, { Component } from "react";
import { Link } from "react-router-dom";
import qs from "qs";
import { Alert } from "@nice-digital/nds-alert";
import { Button } from "@nice-digital/nds-button";
import { Input } from "@nice-digital/nds-input";

import { AuthApi } from "../../services/AuthApi";
import { auth as authOpts } from "../../services/constants";
import { isDomainInUsername, validateLoginFields, scrollToMyRef } from "../../helpers";

import "./Login.scss";

export class Login extends Component {
  constructor(props) {
    super(props);
    this.auth = new AuthApi();
    this.state = {
      username: null,
      password: null,
      clientSideErrors: {
        username: false,
        password: false
      },
      serverSideError: null,
      loading: false,
      isAD: false,
      connection: authOpts.connection,
      showGoogleLogin: false,
      activationEmailSent: false
    };
    this.querystring = qs.parse(document.location.search, {
      ignoreQueryPrefix: true
    });
    this.continue = true

    this.formRefs = Object.keys(this.state.clientSideErrors).reduce((accumulator, error) => {
      return Object.assign(accumulator, {
        [error]: React.createRef()
      });
    }, {});
  }

  componentDidMount() {
    this.querystring = qs.parse(document.location.search, {
      ignoreQueryPrefix: true
    });
    this.auth.fetchClientSettings().then(() => {
      // We are not using google to connect for now
      // this.googleConnection = window.Auth0.strategies && window.Auth0.strategies.google-oauth2 && window.Auth0.strategies.google-oauth2.connectionName || null;

      this.ADConnection = window.Auth0.strategies && window.Auth0.strategies.waad && window.Auth0.strategies.waad.connectionName || null;
      this.setState(
        { showGoogleLogin: !!this.googleConnection },
        this.showAuth0RulesError
      )
    });
    if (this.querystring.register === "true"){
      this.props.history.push("/register");
    }
  }

  showAuth0RulesError = () => {
    this.setState({ serverSideError: this.querystring.myerror })
  };

  resendVerificationEmail = (e) => {
    if (e) e.preventDefault()
    const callback = err =>
      this.setState({ activationEmailSent: !err, serverSideError: err })
    try {
      this.auth.resendVerificationEmail(this.querystring.userid, callback)
    } catch (err) {
      console.log(JSON.stringify(err))
    }
  };

  clientSideHasErrors = (clientSideErrors) =>{
    return clientSideErrors.username || clientSideErrors.password;
  }

  validate = () => {
    const tests = validateLoginFields(this.state);
    const clientSideErrors= {
        username: !this.state.username || tests.username(),
        password: tests.password()
      };
    this.setState({ loading: false, clientSideErrors});
    return !this.clientSideHasErrors(clientSideErrors);
  }

  login = (e, isGoogle) => {
    if (e) e.preventDefault()
    const requestErrorCallback = err =>
      this.setState(
        {
          serverSideError: err.description || err.error_description,
          loading: false
        },
        console.log(JSON.stringify(err))
      );

    if (this.validate()){
      try {
        this.setState({ loading: true, serverSideError: null }, () => {
          const { username, password, connection } = this.state
          const loginConnection = isGoogle ? this.googleConnection : connection
          const isResumingAuthState =
            this.querystring.myerrorcode &&
            this.querystring.myerrorcode === 'user_not_verified'
              ? this.continue && this.querystring.state
              : null
          this.auth.login(
            loginConnection,
            username,
            password,
            requestErrorCallback,
            isResumingAuthState,
            this.props.history || null
          )
        })
      } catch (err) {
        console.log(JSON.stringify(err))
        this.setState({ loading: false, serverSideError: 'Something has gone wrong.' })
      }
    }
  };

  handleChange = ({ target: { name, value } }) => {
    let isAD = null;
    if (name === 'username') {
      isAD = isDomainInUsername(value);
      if (this.querystring.myerrorcode && this.querystring.email !== value) {
        this.continue = false
      }
    }
    this.setState({
      [name]: value,
      isAD,
      connection: isAD ? this.ADConnection : authOpts.connection
    }, () =>{
      if (this.clientSideHasErrors(this.state.clientSideErrors)){
        this.validate();
      }
    });
  };

  render() {
    const {
      serverSideError,
      clientSideErrors,
      loading,
      isAD,
      showGoogleLogin,
      activationEmailSent
    } = this.state

    const showUserNotVerfiedMessage = (this.querystring && this.querystring.myerrorcode === 'user_not_verified');

    const requiredMessage = 'This field is required';

    const errorMessages = {
      username: `Email - ${requiredMessage}`,
      password: `Password - ${requiredMessage}`,
    }

    return (
      <div>
        <h2 className="mt--0">Log in</h2>
        <p className="lead"><Link
          data-qa-sel="Signup-link-login"
          to="/register"
          activeclassname="activeRoute"
        >
          Create a NICE account
        </Link></p>
        <form>
            {(serverSideError || this.clientSideHasErrors(this.state.clientSideErrors)) && (
              <Alert type="error" role="alert" data-qa-sel="problem-alert-login">
                <p className="lead">There is a problem</p>
                <ul>
                  {Object.keys(clientSideErrors).map((errorName, idx) => {
                    if (clientSideErrors[errorName]) {
                      return (
                        <li key={idx}>
                          <a href={`#${errorName}`} onClick={(e) => scrollToMyRef(this.formRefs[errorName], e)} aria-label="Go to error">
                              {errorMessages[errorName]}
                          </a>
                        </li>
                      );
                    }
                    return null;
                  })}
                  {serverSideError && (
                    <li>{serverSideError}</li>
                  )}
                </ul>
                {' '}
                {showUserNotVerfiedMessage ? (
                  <button href="#" onClick={this.resendVerificationEmail}>
                    Resend activation email
                  </button>
                ) : null}
              </Alert>
            )}
            {activationEmailSent && (
              <Alert type="success">An activation email has been sent!</Alert>
            )}
            <Input
              name="username"
              data-qa-sel="login-email"
              label="Email"
              id="username"
              unique="username"
              type="email"
              placeholder="eg: your.name@example.com..."
              onChange={this.handleChange}
              autoComplete="email"
              error={clientSideErrors.username}
              errorMessage={requiredMessage}
              inputRef={this.formRefs['username']}
            />
            {!isAD && (
              <Input
                data-qa-sel="login-password"
                name="password"
                type="password"
                unique="password"
                label="Password"
                onChange={this.handleChange}
                autoComplete="current-password"
                error={clientSideErrors.password}
                errorMessage={requiredMessage}
                inputRef={this.formRefs['password']}
              />
            )}

          {isAD && (
            <Alert type="info">
            When you select sign in you will be directed to another screen where you can sign in through Microsoft using your NICE email address and password.
            </Alert>
          )}

          {!loading ? (
            !showUserNotVerfiedMessage ? (
              <div>
                <Button
                  data-qa-sel="login-button"
                  variant="cta"
                  onClick={e => this.login(e, false)}
                  // disabled={!username}
                >
                  Sign in
                </Button>
                {showGoogleLogin && (
                  <button
                    data-qa-sel="login-button-social"
                    className="iconBtn social"
                    style={{ float: 'right' }}
                    onClick={e => this.login(e, true)}
                  >
                    <span className="buttonLabel">Or </span>
                    <img
                      className="iconBtn-icon"
                      alt="Sign in with google"
                      src={ process.env.VITE_PUBLIC_URL + "/images/btn_google_signin_light_normal_web.png"}
                    />
                  </button>
                )}
              </div>
            ) : (
              // user is not verified. the login button is not going to work.
              <div/>
            )
          ) : (
            'Loading...'
          )}
        </form>

        {!isAD && (
        <Link className="forgotPasswordLink" data-qa-sel="forgotPassword-link" to="/forgotPassword">
          Forgot password?
        </Link>
        )}
      </div>
    )
  }
}
