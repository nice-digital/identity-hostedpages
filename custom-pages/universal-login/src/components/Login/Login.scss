.iconBtn {
  border:none;
  background: none;
}
.iconBtn-icon:hover {
  opacity: 0.8;
}
.buttonLabel {
  font-weight: 700;
  padding-right: 6px;
}
