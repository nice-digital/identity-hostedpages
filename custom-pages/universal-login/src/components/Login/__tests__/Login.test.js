import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { Login } from "../Login";

jest.mock("../../../helpers/isLoginPage", () => true);
jest.mock("../../../services/AuthApi", () => ({
	...jest.requireActual("../../../services/AuthApi"),
	fetchClientSettings: jest.fn()
}));

describe("Login components", () => {
  window.Auth0 = {
    strategies: {
      waad: {
        domainString: "gotham.com"
      },
      "google-oauth2": {
        connectionName: "google"
      }
    }
  };

  it("should render Login correctly", () => {
		const { container } = render(
			<MemoryRouter future={{ v7_relativeSplatPath: true, v7_startTransition: true }}>
				<Login />
			</MemoryRouter>
		);
    expect(container).toMatchSnapshot();
  });

//   it('should render Login correctly', () => {
//     expect(el).toMatchSnapshot()
//   })

//   it('should instantiate AuthApi', () => {
//     expect(instance.auth).toBe(auth)
//   })

//   it('should render correctly when there is an error', () => {
//     el.setState({ error: 'this is an error' }).update()
//     expect(el).toMatchSnapshot()
//   })

//   it('should render correctly when there is username and password', () => {
//     el.setState({
//       valid: true
//     }).update()
//     expect(el).toMatchSnapshot()
//   })

//   it('should call the AuthApi login when login is invoked', () => {
//     const username = 'username'
//     const password = 'pwd'
//     const { connection } = authOpts

//     el.setState({
//       username,
//       password
//     }).update()

//     instance.login()
//     expect(instance.auth.login).toBeCalledWith(
//       connection,
//       username,
//       password,
//       expect.any(Function),
//       null,
//       null
//     )
//   })

//   it('clientSideHasErrors validates the username properly', () => {
//     expect(!!instance.clientSideHasErrors({ username: null, password: null })).toBe(false);
//     expect(!!instance.clientSideHasErrors({ username: false, password: false })).toBe(false);
//     expect(!!instance.clientSideHasErrors({ username: true, password: true })).toBe(true);
//     expect(!!instance.clientSideHasErrors({ username: true, password: false })).toBe(true);
//     expect(!!instance.clientSideHasErrors({ username: false, password: true })).toBe(true);
//   })

//   it('should show client side errors when the username or password have not been supplied', () => {
//     const stateBefore = el.state();
//     expect(instance.clientSideHasErrors(stateBefore.clientSideErrors)).toBe(false);
//     expect(stateBefore.username).toBeNull();
//     expect(stateBefore.password).toBeNull();
//     expect(stateBefore.clientSideErrors.username).toBe(false);
//     expect(stateBefore.clientSideErrors.password).toBe(false);

//     const loginButton = el.find("button[data-qa-sel='login-button']");
//     loginButton.simulate("click");

//     const stateAfter = el.state();
//     expect(instance.clientSideHasErrors(stateAfter.clientSideErrors)).toBe(true);
//     expect(stateBefore.username).toBeNull();
//     expect(stateBefore.password).toBeNull();
//     expect(stateAfter.clientSideErrors.username).toBe(true);
//     expect(stateAfter.clientSideErrors.password).toBe(true);
//   })

// })

  it("should render correctly when there is an error", async () => {
		window.HTMLElement.prototype.scrollIntoView = function() {};
		render(
			<MemoryRouter future={{ v7_relativeSplatPath: true, v7_startTransition: true }}>
				<Login />
			</MemoryRouter>
		);
		const loginButton = await screen.findByRole("button", { name: "Sign in"});
		userEvent.click(loginButton);
		const errorAlert = await screen.findByRole("alert");
  	expect(errorAlert).toBeInTheDocument();
  });
});
