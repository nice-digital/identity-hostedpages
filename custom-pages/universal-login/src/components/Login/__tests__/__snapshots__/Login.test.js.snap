// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login components should render Login correctly 1`] = `
<div>
  <div>
    <h2
      class="mt--0"
    >
      Log in
    </h2>
    <p
      class="lead"
    >
      <a
        activeclassname="activeRoute"
        data-qa-sel="Signup-link-login"
        href="/register"
      >
        Create a NICE account
      </a>
    </p>
    <form>
      <div
        class="input"
        data-component="input"
      >
        <label
          class="input__label"
          for="username"
        >
          Email
        </label>
        <input
          autocomplete="email"
          class="input__input"
          data-qa-sel="login-email"
          id="username"
          name="username"
          placeholder="eg: your.name@example.com..."
          type="email"
          unique="username"
          value=""
        />
      </div>
      <div
        class="input"
        data-component="input"
      >
        <label
          class="input__label"
          for="password"
        >
          Password
        </label>
        <input
          autocomplete="current-password"
          class="input__input"
          data-qa-sel="login-password"
          id="password"
          name="password"
          type="password"
          unique="password"
          value=""
        />
      </div>
      <div>
        <button
          class="btn btn--cta"
          data-component="button--cta"
          data-qa-sel="login-button"
          type="button"
        >
          Sign in
        </button>
      </div>
    </form>
    <a
      class="forgotPasswordLink"
      data-qa-sel="forgotPassword-link"
      href="/forgotPassword"
    >
      Forgot password?
    </a>
  </div>
</div>
`;
